plugins/modules/foreman_hostgroup.py

```diff
@@ -428,14 +428,21 @@ def main():
             current_puppetclasses = [p['id'] for p in current_puppetclasses]
             for puppet_class_name in puppetclasses:
                 scope = {'environment_id': hostgroup['environment_id']}
-                results = module.list_resource("puppetclasses", params=scope, search='name="{0}"'.format(puppet_class_name))
-                if len(results) == 1 and len(list(results.values())[0]) == 1:
-                    puppet_class_id = list(results.values())[0][0]['id']
+
+                # puppet classes API return puppet classes grouped by puppet module name
+                puppet_modules = module.list_resource("puppetclasses", params=scope, search='name="{0}"'.format(puppet_class_name))
+                puppet_classes = list(puppet_modules.values())[0]
+
+                # verify that only one puppet module is returned with only one puppet class inside
+                # as provided search results have to be like "results": { "ntp": [{"id": 1, "name": "ntp" ...}]}
+                # and get the puppet class id
+                if len(puppet_modules) == 1 and len(puppet_classes) == 1:
```

Comment:
That's a lot more readable! But the `len(puppet_modules) == 1` test must be before `puppet_classes = list(puppet_modules.values())[0]` imho.
I think, before it worked due to lazy evaluation of boolean expressions.
I'd say it would be ok to first do the combined check, and then break out the `puppet_classes` in a variable.

Reply:
This line can fail if puppet class is not found - so it will fail a bit earlier with a less pretty error message. I moved the var assignment as you proposed.